File: Solved problems/kattis/AddingWords.cpp
// tags: implementation
#include <bits/stdc++.h>
using namespace std;
// -------------------------------
map<string,int> word2val;
map<int,string> val2word;
int main() {
    string line;
    while (getline(cin, line)) {        
        stringstream ss;
        ss << line;
        string command;
        ss >> command;
        if (command == "clear") {
            word2val.clear();
            val2word.clear();
        } else if (command == "def") {
            string w;
            int v;
            ss >> w >> v;
            if (word2val.count(w)) {
                val2word.erase(word2val[w]);
            }
            word2val[w] = v;
            val2word[v] = w;
        } else {
            assert (command == "calc"); // paranoico
            string w, op;
            ss >> w;
            int ans = 0;
            bool valid = true;
            if (word2val.count(w) == 0) { // not found
                valid = false;
            } else {
                ans = word2val[w];
            }
            while (valid and (ss >> op >> w)) {
                if (word2val.count(w) == 0) {// not found
                    valid = false;
                    break;
                }
                if (op == "+") ans += word2val[w];
                else ans -= word2val[w];
            }
            cout << line.substr(5) << " ";
            if (valid) {
                auto it = val2word.find(ans);
                if (it == val2word.end()) {
                    cout << "unknown\n";                    
                } else {
                    cout << it->second << '\n';
                }
            } else {
                cout << "unknown\n";
            }
        }
    }
    return 0;
}

File: Solved problems/kattis/AddingWords.py
import sys

def main():
    word2val = dict()
    val2word = dict()
    for line in sys.stdin:
        line = line.strip()
        tokens = line.split()
        command = tokens[0]
        if command == 'clear':
            word2val.clear()
            val2word.clear()
        elif command == 'def':
            w = tokens[1] # word
            v = int(tokens[2]) # value
            if w in word2val: # caso borde
                old_v = word2val[w]
                # del word2val[w] # overkill, pero por si las moscas
                del val2word[old_v] # sí o sí
                # del val
            word2val[w] = v
            val2word[v] = w
        else:
            assert command == 'calc' # paranoico
            try:
                res = word2val[tokens[1]]
                for i in range(2, len(tokens)-1, 2):
                    if (tokens[i] == '+'):
                        res += word2val[tokens[i+1]]
                    else:
                        res -= word2val[tokens[i+1]]
                res = val2word[res]
            except KeyError:
                res = 'unknown'
            print(line[5:], res)

main()